Board category parse desc

MissAllSunday opened the pull request:
Fixes #5734

Made a refactor on how to parse and save the parsed content on the cache.

Its a somehow atomic approach, the get and set operations are done on a per category and per board basis, this means an user who can only see a subset of the total boards won't get the load of parsing all of them.

Tested with filebased and SQLite3

MissAllSunday committed:
getParsedDescriptions()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
Quid pro quo Clarice...

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
setParsedDescriptions()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
setParsedDescriptions()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
shenanigans

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday commented:
Ugh, I forgot to replace the data with its parsed counterpart. Its late so I'll do it later.

MissAllSunday committed:
Forgot about child boards

Signed-off-by: Jessica González <suki@missallsunday.com>

albertlast commented:
> Some poor soul will get the hit of parsing the entire categories and boards names and descriptions but there isn't much we can do about it.

How would be to allow to disable this feature like "recent post function"?

And how long run you change on ~100 board forum in compare to the existing one?

MissAllSunday commented:
I don't understand what you mean by disable this feature. No it cannot be disable unless you don't use cache at all.

You won't be able to measure this against other alternatives. This approach adds to the cache as needed.  The query uses {query_see_board} which means the load of parsing the names and description will be shared among several users since all of them will have access to different boards.

Every time there's something new to parse, the cache gets refreshed and the new data gets appended to what we already have. The cache also has a long time duration since boards and categories aren't updated that frequently.

Lastly, having the cache by category means we will end up with fewer cache entries of a moderate size.

I also need to re-create the cache whenever a board is created/updated. Will add that too.

albertlast commented:
I don't see the bcode parsing feature in general as essential,
so i would purpose to allow to disable the parsing function to safe runtime/cpu time.

I do all the time measurements, first alternative is the version before you add you first pr,
the secodn one is the merge version right now.

for both cases: with cache and none cache.

MissAllSunday commented:
> I do all the time measurements, first alternative is the version before you add you first pr,
the secodn one is the merge version right now.

Cool. What I'm saying is that if you try to measure this using an admin account who can see 100 boards then you won't be able to see much improvements.

albertlast commented:
sm.org would already reach the point and
i'm waiting for measurments...

MissAllSunday commented:
> sm.org would already reach the point and

???

MissAllSunday committed:
appendCategoriesParsedDescriptions();
appendBoardsParsedDescriptions()

Signed-off-by: Jessica González <suki@missallsunday.com>

albertlast commented on Sources/Subs-BoardIndex.php:
```diff
@@ -95,7 +95,7 @@ function getBoardIndex($boardIndexOptions)
 		$result_boards = $smcFunc['db_query']('', '
 			SELECT' . ($boardIndexOptions['include_categories'] ? '
-				c.id_cat, c.name AS cat_name, c.description AS cat_desc,' : '') . '
+				c.id_cat, c.name AS cat_name, c.description AS cat_desc,' : 'b.id_cat,') . '
```
i would also then change the c.id_cat to b.id_cat in the first position

albertlast reviewed (commented):
only look throw the code

albertlast commented:
While testing i get two kind of errors:
```
Error

    Type of error: Undefined_vars
    Error message: 8: Undefined index: description
    File: F:/github/SMF2.1/Sources/Subs-BoardIndex.php
    Line: 621
    URL of page causing the error: http://localhost/SMF2.1/index.php?

Backtrace information

    #0: smf_error_handler()
    Called from F:\github\SMF2.1\Sources\Subs-BoardIndex.php on line 621
    #1: setParsedDescriptions()
    Called from F:\github\SMF2.1\Sources\Subs-BoardIndex.php on line 548
    #2: getBoardIndex()
    Called from F:\github\SMF2.1\Sources\BoardIndex.php on line 49
    #3: BoardIndex()
    Called from F:\github\SMF2.1\index.php on line 155

```

and
```
Error

    Type of error: Undefined_vars
    Error message: 8: Undefined index: name
    File: F:/github/SMF2.1/Sources/Subs-BoardIndex.php
    Line: 620
    URL of page causing the error: http://localhost/SMF2.1/index.php?

Backtrace information

    #0: smf_error_handler()
    Called from F:\github\SMF2.1\Sources\Subs-BoardIndex.php on line 620
    #1: setParsedDescriptions()
    Called from F:\github\SMF2.1\Sources\Subs-BoardIndex.php on line 548
    #2: getBoardIndex()
    Called from F:\github\SMF2.1\Sources\BoardIndex.php on line 49
    #3: BoardIndex()
    Called from F:\github\SMF2.1\index.php on line 155
```

with this i can't test you changes,
only when i disable cache(used APCu) i can run your code fine,
ther i notice that your code run slower as the existing solution.

MissAllSunday committed:
Parse and cache the new content

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
Parse and cache the new category content

Signed-off-by: Jessica González <suki@missallsunday.com>

live627 commented on Sources/Subs-BoardIndex.php:
```diff
@@ -470,13 +544,141 @@ function getBoardIndex($boardIndexOptions)
 	}
 
+	// I took my time, I hurried up, the choice was mine I didn't think enough
+	$parsed_descriptions = setparsedDescriptions($to_parse);
```
Just a tiny nitpick here... sorry... case doesn't match

MissAllSunday replied:
case?

albertlast replied:
i guess it should be setParsedDescriptions

MissAllSunday committed:
fix: Add cache only on updates

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: fix: check $category['boards']

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
blank canvas

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
blank canvas

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: setCategoryParsedDescriptions()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: getCategoryParsedDescription()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
feat: setBoardParsedDescription()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
feat: getBoardsParsedDescription()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
feat: getBoardsParsedDescriptionById()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
feat: apply setBoardParsedDescription on create and modify boards

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: category names don't get parsed

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
feat: get the category and board parsed info on display

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: return $parsed_board_info for board info

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: travis complains

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday commented:
Made a refactor on how to parse and save the parsed content on the cache.

Its a somehow atomic approach, the get and set operations are done on a per category and per board basis, this means an user who can only see a subset of the total boards won't get the load of parsing all of them.

Tested with filebased and SQLite3.

Please give it a test.

albertlast commented:
When you switch from caching back to non caching,
stuff get lost:
![grafik](https://user-images.githubusercontent.com/1782906/75104424-22fe3180-5609-11ea-92a7-47f2b198e5fb.png)

Since this bug i can't fully do a performance test,
but partial:
You new version is 10ms faster as the existing code in a cached env.
Non cached env is missing of this bug mention of the start

MissAllSunday committed:
fix: use a single cache entry for category descriptions

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: cache the result if the setting is on

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: cache the result if the setting is on

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: set bulk cache entries

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: set bulk cache entries on board create/modify

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: Some boards didn't get their info cached, it is done on a per category basis

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: replace the unparsed data for boards with no cache entry

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: remove unused function getBoardParsedDescriptionById()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday commented:
Added yet another refactor, grouping both categories and board caches, now there's a single cache entry for all categories and board cache is grouped by category, reducing the amount of hits to get the cached data.

MissAllSunday committed:
fix: move b.id_cat to $board_index_selects

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: declare $boards_parsed_data for both parent and child boards

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: add support for child boards parsed descriptions

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: re-parsing missing child board descriptions

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: $category_description defaults to $row_board['cat_desc'];

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: Add support for parsing board name and desc on MessageIndex
fixes #6307

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: move call to setBoardParsedDescription() after clean_cache call

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: Avoid losing already parsed data due to call to clean_cache()

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: drop support for parsing board name

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday committed:
fix: do not cache empty values

Signed-off-by: Jessica González <suki@missallsunday.com>

MissAllSunday commented:
I removed support for parsing the board/cat name its just too much hassle to make sure the parsed result won't mess up all the different places where the board name is displayed.

Also due to the call to clean_cache when a board is been created/modified all cache data will be lost, I added support for keeping the cache data for all the other boards on the same category but the rest will get cleaned.

MissAllSunday committed:
fix: spaces to tabs

Signed-off-by: Jessica González <suki@missallsunday.com>
